manuals/sccainfo.1
.Dd June 25, 2026
.Dt SCCAINFO 1
.Os
.Sh NAME
.Nm sccainfo
.Nd determines information about a Windows Prefetch File (PF)
.Sh SYNOPSIS
.Nm sccainfo
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm sccainfo
is a utility to determine information about a Windows Prefetch File (PF)
.Pp
.Nm sccainfo
is part of the
.Nm libscca
package.
.Nm libscca
is a library to access the Windows Prefetch File (PF) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# sccainfo CMD.EXE-4A81B364.pf
sccainfo 20110704

Windows Prefetch File (PF) information:
	...

.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libscca/issues
.Sh COPYRIGHT
Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* move directory strings to a separate structure without the overhead of libfvalue_value_append_entry_data
* add API functions
  - get file size ?
  - file metrics
    - get filename offset?
  - volume information
    - get directory strings
  - add functions to calcalate hashes?
* handle directory names correctly
* documentation
  - describe relation ship metrics and trace chains
* update manuals

20141003
* see `git log' for more recent change log
* removed README.macosx
* added dpkg files
* changes for project site move

20140629
* Worked on format support

20140614
* Worked on format support

20140423
* worked on API functions

20140421
* Worked on format support

20140412
* Worked on format support

20140411
* Worked on format support

20140410
* worked on API functions

20140402
* updated dependencies

20140324
* 2014 update
* updated dependencies

20131020
* renamed to libscca

20130922
* updated dependencies

20130609
* 2013 update
* updated dependencies

20121112
* 2012 update
* updated dependencies
* code clean up
* updated pc file
* updated spec file

20111113
* worked on initial version

20111112
* Initial version based on libexe



sccatools/sccainfo.c
/*
 * Shows information obtained from a Windows Prefetch File (PF).
 *
 * Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "sccatools_getopt.h"
#include "sccatools_libcerror.h"
#include "sccatools_libclocale.h"
#include "sccatools_libcnotify.h"
#include "sccatools_libscca.h"
#include "sccatools_output.h"
#include "sccatools_signal.h"
#include "sccatools_unused.h"

info_handle_t *sccainfo_info_handle = NULL;
int sccainfo_abort                  = 0;

/* Signal handler for sccainfo
 */
void sccainfo_signal_handler(
      sccatools_signal_t signal SCCATOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "sccainfo_signal_handler";

	SCCATOOLS_UNREFERENCED_PARAMETER( signal )

	sccainfo_abort = 1;

	if( sccainfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     sccainfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use sccainfo to determine information about a Windows Prefetch File (PF).";

	sccatools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libscca_error_t *error     = NULL;
	system_character_t *source = NULL;
	char *program              = "sccainfo";
	system_integer_t option    = 0;
	int number_of_options      = (int) ( sizeof( options ) / sizeof( sccatools_option_t ) );
	int verbose                = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "sccatools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( sccatools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	sccatools_output_version_fprint(
	 stdout,
	 program );

	if( sccatools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = sccatools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				sccatools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				sccatools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				sccatools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		sccatools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libscca_notify_set_stream(
	 stderr,
	 NULL );
	libscca_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &sccainfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( info_handle_open_input(
	     sccainfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source file.\n" );

		goto on_error;
	}
	if( info_handle_file_fprint(
	     sccainfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print file information.\n" );

		goto on_error;
	}
	if( info_handle_close_input(
	     sccainfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &sccainfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( sccainfo_info_handle != NULL )
	{
		info_handle_free(
		 &sccainfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

